tests/loggen/file_reader.c

```diff
@@ -103,79 +104,234 @@ void close_file_reader(int nr_threads)
+  safe_copy_string(elements->sdata, "", PARS_BUF_SDATA_SIZE, "sdata"); /* no sdata in RFC3164 */
+
+  /* skip the leading '<' char */
+  if (g_str_has_prefix(line, "<"))
+    line = line + 1;
+
+  gchar *pri_end = g_strstr_len(line, PARS_BUF_MSG_SIZE, ">");
+  if (!pri_end)
+    {
+      ERROR("Invalid rfc3164 log format. No '>' found.\n");
+      return -1;
     }
+
+  /* pri + time stamp */
+  safe_copy_n_string(elements->pri, line, pri_end - line, PARS_BUF_PRI_SIZE, "pri");
+  safe_copy_n_string(elements->time_stamp, pri_end+1, RFC3164_TIMESTAMP_SIZE, PARS_BUF_TIME_STAMP_SIZE, "timestamp");
+  gchar *timestamp_end = pri_end + 1 + RFC3164_TIMESTAMP_SIZE;
+
+  /* host name */
+  gchar *host_begin = timestamp_end + 1;
+  gchar *host_end = g_strstr_len(host_begin, PARS_BUF_MSG_SIZE, " ");
+  if (host_end)
+    safe_copy_n_string(elements->host, host_begin, host_end - host_begin, PARS_BUF_HOST_SIZE, "host");
   else
     {
-      pid[0] = '\0';
-      ++pos; /* the 'pos' has been decreased in the condition (']'), reset it to the original position */
-    }
+      safe_copy_string(elements->host, host_begin, PARS_BUF_HOST_SIZE, "host");
+      safe_copy_string(elements->pid, "", PARS_BUF_PID_SIZE, "pid");
+      safe_copy_string(elements->app, "", PARS_BUF_APP_SIZE, "app");
+      safe_copy_string(elements->message, "", PARS_BUF_MSG_SIZE, "msg");
 
-  /* Program */
-  end = pos;
-  while (*(--pos) != ' ')
-    ;
+      return 1;
+    }
 
-  memcpy(program, pos + 1, end - pos - 1);
-  program[end-pos-1] = '\0';
+  /* application[pid]: */
+  gchar *app_begin = host_end + 1;
+  gchar *app_end = g_strstr_len(app_begin, PARS_BUF_MSG_SIZE, " ");
 
-  /* Host */
-  end = pos;
-  while (*(--pos) != ' ')
-    ;
+  /* netither PID nor application found */
+  if ( !g_strstr_len(app_begin, PARS_BUF_APP_SIZE, ":" ) )
+    {
+      safe_copy_string(elements->pid, "", PARS_BUF_PID_SIZE, "pid");
+      safe_copy_string(elements->app, "", PARS_BUF_APP_SIZE, "app");
+      safe_copy_string(elements->message, app_begin, PARS_BUF_MSG_SIZE, "msg");
+      return 1;
+    }
 
-  memcpy(host, pos + 1, end - pos - 1);
-  host[end-pos-1] = '\0';
+  gchar *pid_begin = g_strstr_len(app_begin, PARS_BUF_APP_SIZE, "[");
+  gchar *pid_end   = g_strstr_len(app_begin, PARS_BUF_APP_SIZE, "]");
+  if ( !pid_begin || !pid_end )
+    {
+      safe_copy_string(elements->pid, "", PARS_BUF_PID_SIZE, "pid");
+      safe_copy_n_string(elements->app, app_begin, app_end - app_begin - 1, PARS_BUF_APP_SIZE, "app");
+    }
+  else
+    {
+      safe_copy_n_string(elements->pid, pid_begin + 1, pid_end - pid_begin - 1, PARS_BUF_PID_SIZE, "pid");
+      safe_copy_n_string(elements->app, app_begin, pid_begin - app_begin, PARS_BUF_APP_SIZE, "app");
+    }
 
-  *msg = ((char *)pos0) + 2;
+  safe_copy_string(elements->message, app_end + 1, PARS_BUF_MSG_SIZE, "msg");
 
   return 1;
 }
 
+static int
+parse_line(const char *line, SyslogMsgElements *elements, int skip)
```

Comment:
Create a separate `skip_tokens` function(and maybe we could call this function before `parse_line`).